### PR TITLE
Hiding tooltips on scroll

### PR DESCRIPTION
**Pull request by herschel666**
I experienced the tooltip to stay visible, during scrolling (FF 39, MacBook Pro OS X 10.10.3, touchpad). To fix this, I added an event-listener & hide the tooltip on scrolling.

### CODE REVIEW

**Comment by Garbee** [comment]
So fix for issue #545. Assigning to Surma for review since the main issue is assigned to him.

**Comment by surma** [comment]
I had the same solution in mind, but I am hesitant to merge. Every menu component on a page will register a scroll handler that will get called on every scroll event. I feel like this is will be a performance issue. That’s why #545 is still open.

@addyosmani @sgomes ?

**Comment by Garbee** [comment]
cc @samccone for some perf ideas.

**Comment by samccone** [comment]
>  Every menu component on a page will register a scroll handler that will get called on every scroll event

Unless I am reading this wrong (first time looking at the time) This should only happen if a tooltip has been clicked and opened though right?

**Comment by surma** [comment]
Yes, it’s mostly an issue on mobile, where you need to tap to show the tooltip.

**Comment by samccone** [comment]
So hypothetically lets assume a user on mobile figures out a way to open 100 tooltips. (not sure how they would do this but lets assume this magic world exists)

They scroll 1px and our 100 event registrations fire. 
The one thing that I think **might** cause a perf issue (but hard to say for sure until it is perfed) would be this line

```js
this.element_.classList.remove(this.CssClasses_.IS_ACTIVE);
```

Since it will cause a repaint (but not a page reflow or anything) it might possibility bust our 16ms frame causing scroll jank :hankey: .

What can we do...
Wrap it in a RAF?  Throttle the scroll callbacks?


thoughts @surma ?
This entire argument though is predicated on the belief that this will cause a perf hit which I am just not sure about yet.

**Comment by surma** [comment]
> Throttle the scroll callbacks?

I was thinking about debouncing them as well somehow.

But I guess you are right, maybe there is no perf issue after all and we should just merge this.

**Comment by Garbee** [comment]
I'd say merge it and if there is a perf issue fix it later. This at least helps address the immediate problem now.

**Comment by surma** [comment]
Sadly, this PR does not fix the issue on the components page for me. I’ll investigate.

**Comment by herschel666** [comment]
Listening for `touchmove` — as well — worked for me. Any oppinions on that?

Btw: the CI-testing instantly failed, though locally everything worked fine for me with `gulp`.

Cheers.

**Comment by addyosmani** [comment]
Sorry about the CI failures. We're looking into what's going on with Drone (cc @crhym3)

**Comment by surma** [comment]
`touchmove` solves the problem! Good idea, @herschel666.

File: src/tooltip/tooltip.js [hunk 1]
```diff
@@ -64,6 +64,8 @@ MaterialTooltip.prototype.handleMouseEnter_ = function(event) {
   this.element_.style.marginLeft = -1 * (this.element_.offsetWidth / 2) + 'px';
   this.element_.style.top = props.top + props.height + 10 + 'px';
   this.element_.classList.add(this.CssClasses_.IS_ACTIVE);
+  window.addEventListener('scroll', this.boundMouseLeaveHandler, false);
+  window.addEventListener('touchmove', this.boundMouseLeaveHandler, false);
```
Comment (surma) [inline src/tooltip/tooltip.js]: Would you mind moving these to the `init()` like all the other event handlers?

File: src/tooltip/tooltip.js [hunk 2]
```diff
@@ -76,6 +78,8 @@ MaterialTooltip.prototype.handleMouseLeave_ = function(event) {
 
   event.stopPropagation();
   this.element_.classList.remove(this.CssClasses_.IS_ACTIVE);
+  window.removeEventListener('scroll', this.boundMouseLeaveHandler);
+  window.removeEventListener('touchmove', this.boundMouseLeaveHandler, false);
```
Comment (surma) [inline src/tooltip/tooltip.js]: Would you mind moving these to the `mdlDowngrade_()` like all the other event handlers?

Comment (Garbee) [inline src/tooltip/tooltip.js]: Move > Copy.

Moving would mean breaking the fix itself. Copy will simply ensure that they are removed when downgraded in-case missed at some other point.

File: src/tooltip/tooltip.js (hunk 1 above)
Comment (surma) [inline src/tooltip/tooltip.js]: Never mind this! Only attaching these handlers when inside the element is actually a good idea, imo.

File: src/tooltip/tooltip.js (hunk 2 above)
Comment (surma) [inline src/tooltip/tooltip.js]: Nvm. See above.

**Comment by surma** [comment]
LGTM. Second pair of eyes please, @sgomes

**Comment by sgomes** [comment]
Code-wise looks good, @surma. Need me to check how it runs?

**Comment by surma** [comment]
@sgomes Nah, just wanted to have someone look at the code. I tested it. Thanks for looking.  Merging.

**Comment by surma** [comment]
@herschel666 Thanks a lot for your work!

**Comment by samccone** [comment]
:boom:

**Comment by herschel666** [comment]
You're welcome! :-)

**Pull request merged**